lib/Lexer/PHP.hs

```diff
@@ -0,0 +1,121 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE QuasiQuotes, OverloadedStrings #-}
+module Lexer.PHP() where
+
+import qualified Data.List.NonEmpty as NE
+import Regex.Parse
+import Data.Maybe
+import Data.Text (Text)
+import qualified Data.Text as T
+import Lexer.Types
+import Lang
+import Utils
+
+instance LexerWriter PHP where
+  writeLexer _ accSt tokNames stList =
+    [ ("Lexer.php", [interp|
+<?php
+
+namespace Alpaca;
+
+class Lexer
+{
+    const TOKEN_TYPE_EOF = 0;
+    #{indent 1 tokDefns}
+
+    private $input;
+    private $curChIx;
+    private $debug;
+
+    public function __construct(string $input, bool $debug)
+    {
+        $this->input = $input;
+        $this->curChIx = 0;
+        $this->debug = $debug;
+    }
+
+    public static function tokenToString(int $token): string
+    {
+        switch($token) {
+            case 0: return '%eof';
+            #{indent 3 tokToStr}
+            default: throw new \\InvalidArgumentException("Unknown token $token");
+        }
+    }
+
+    public function getNextToken(): array
+    {
+        $lastAccChIx = $this->curChIx;
+        $startChIx = $this->curChIx;
+        $accSt = -1;
+        $curSt = 0;
+        while ($curSt >= 0) {
+            if (in_array($curSt, [#{T.intercalate "," $ map (tshow . fst) accSt}])) {
+                $lastAccChIx = $this->curChIx;
+                $accSt = $curSt;
+            }
+
+            if ($this->curChIx >= strlen($this->input))
+                break;
+
+            $curCh = $this->input[$this->curChIx];
+            $this->curChIx+=1;
+            switch($curSt) {
+                #{indent 4 transTable}
+            }
+            break;
+        }
+
+        $lastReadChIx = $this->curChIx;
+        $this->curChIx = $lastAccChIx;
+        $text = substr($this->input, $startChIx, $lastAccChIx - $startChIx);
+        switch($accSt) {
+            #{indent 3 returnResult}
+        }
+
+        if ($this->curChIx >= strlen($this->input)) {
+            if ($this->debug)  printf("Got EOF while lexing \\"%s\\"\\n", $text);
+            return [self::TOKEN_TYPE_EOF, null];
+        }
+        throw new \\InvalidArgumentException("Unexpected input: " . substr($this->input, $startChIx, $lastReadChIx));
+    }
+}
+|])]
+    where
+    indent = indentLang 4
+    returnResult = T.intercalate "\n" (map returnResult1 accSt)
+    returnResult1 :: (Int, (Maybe Text, Action)) -> Text
+    returnResult1 (st, (Just name, act)) = [interp|
+        case #{st}:
+            if ($this->debug) printf('Lexed token #{name}: "%s"', $text);
```

Comment:
There are two more places with `printf`. Apparently it is was not obvious I was talking about all of them. These also need newlines...